### ISSUE TITLE
Built-in Messaging API

### ISSUE THREAD

**Issue by aklinker1** [wxt-dev/wxt#643]
### Feature Request

Replacement for `browser.runtime.sendMessage`/`browser.runtime.connect`/`browser.tabs.sendMessage`. The APIs provided by Chrome are difficult to use, not type-safe, and leads to bad UX.

In particular, the messaging API should support the following:

- [Internal messaging](https://developer.chrome.com/docs/extensions/develop/concepts/messaging#simple)
- [Webpage messaging](https://developer.chrome.com/docs/extensions/develop/concepts/messaging#external-webpage)
- [Native messaging](https://developer.chrome.com/docs/extensions/develop/concepts/native-messaging)
- [User scripts messaging](https://developer.chrome.com/docs/extensions/reference/api/userScripts#messaging)
- Messaging between injected scripts and content scripts

Additionally, there are multiple ways of architecting an extension around storage/messaging. I don't think a single solution will work for all these architechures. We may want to consider different APIs for each?

| Background as an "API" | Event-based | Decentralized |
| --- | --- | --- |
| <img width="360" src="https://github.com/wxt-dev/wxt/assets/10101283/5f61bcc6-5526-4ea5-a567-b8f9accac1df"> | <img width="360" src="https://github.com/wxt-dev/wxt/assets/10101283/4a84a915-a684-4c11-9e96-e5f9d9704ade"> | <img width="360" src="https://github.com/wxt-dev/wxt/assets/10101283/8c568062-ec5c-4e89-9a4a-b1dbd733c7b7"> |
| `import { ... } from 'wxt/rpc'` |  `import { ... } from 'wxt/events'` |  `import { ... } from 'wxt/messaging'` | 

> These are all the main ways I'm aware of at least.

That said, I don't like the decentralized approach, and would prefer to omit it entirely from WXT. It basically turns into the "background as an API" in the end anyways, since messages have to be passed through the background.

#### Is your feature request related to a bug?

N/A

### What are the alternatives?

Don't include a messaging API in WXT, and have people use their third-party package of choice (what we do now).

**Comment by manojhl** [comment]
I believe having a wxt way of sending Messaging API is critical, maybe create a separate package wxt-messaging to keep it as optional.

I come from plasmo ecosystem, though i use both @plasmohq/messaging and webext-bridge. I like webext-bridge way of doing things, it's simple but limited.

**Comment by aklinker1** [comment]
@manojhl thanks for the feedback! I never used plasmo's messaging APIs, I'll look at their API for reference as well.

Eventually, I plan on publishing all these extra utils (including storage and localization) as separate packages so you can use them outside WXT as well.

**Comment by spookyuser** [comment]
I also like background as an api! I don't love plasmo's messaging because there's so much setup involved to create a new api function vs trpc which is just one line. 

I too have been enjoying https://github.com/tshaddix/webext-redux and (ofc) https://webext-core.aklinker1.io/guide/proxy-service, being able to switch between contexts makes the messaging apis feel particularly annoying soo I also like the proxying approach with everything in the bg.

Also https://github.com/janek26/trpc-browser seems to be better maintained, but the last time I tried it I kept getting error messages related to chrome ports so I haven't used it since, but I think these may have actually been because of plasmo 😬

I don't understand react server components nearly well enough to know if this is even conceptually possibly, but it would be cool if it were possible to run the react-dom server in the bg and use it as a react server for components in the content script

**Comment by aklinker1** [comment]
> I don't understand react server components nearly well enough to know if this is even conceptually possibly, but it would be cool if it were possible to run the react-dom server in the bg and use it as a react server for components in the content script

Man, I have no idea how this would work lol. Maybe with offscreen documents? But I guess I don't understand why it would be useful to "server render" UI in an extension... But an interesting idea non-the-less.

> I too have been enjoying https://github.com/sinanbekar/webext-zustand 

Didn't know about this one, thanks for sharing!

**Comment by spookyuser** [comment]
> Man, I have no idea how this would work lol. Maybe with offscreen documents? But I guess I don't understand why it would be useful to "server render" UI in an extension... But an interesting idea non-the-less.

lol me too - I think the only reason would be being able to manage all the state in one place and being able to do fetch requests and stuff without having to proxy them through the background first so I guess it would remove the need for the trpc-browser layer / js object proxy

**Comment by axuj** [comment]
@webext-core/proxy-service works very well, but after build, both ui and background will have called code. I wrote a lib that ui only use typescript, and support pass stream data in asyncGeneratorFunctions.

**Comment by aklinker1** [comment]
> @webext-core/proxy-service works very well, but after build, both ui and background will have called code

Yeah, that's something I've known about for a while, but have never fixed. Would require a breaking change to fix.

Instead, I've been using [`trpc-chrome`](https://github.com/jlalmes/trpc-chrome) for my communication needs. It's a little more setup, but I think it creates a more unified API and includes tons of other features like validation.

**Comment by axuj** [comment]
> > @webext-core/proxy-service works very well, but after build, both ui and background will have called code
> 
> Yeah, that's something I've known about for a while, but have never fixed. Would require a breaking change to fix.
> 
> Instead, I've been using [`trpc-chrome`](https://github.com/jlalmes/trpc-chrome) for my communication needs. It's a little more setup, but I think it creates a more unified API and includes tons of other features like validation.

Validation is not as important in browser extensions as it is in web APIs, in my opinion. You can check out my library [webext-rpc](https://github.com/axuj/webext-rpc) which works like @webext-core/proxy-service but doesn't build the dependency into the UI.

**Comment by aklinker1** [comment]
I like your library a lot! Like the API, thanks for sharing!

**Comment by manojhl** [comment]
Just mentioning @extend-chrome's https://github.com/extend-chrome/messages package which is built on promises and observables.

**Comment by Timeraa** [comment]
Just gonna comment on this, I'm currently trying my best to implement this in a good type-safe way, so far it's looking good, taking a lot of inspiration from @aklinker1's Proxy service but supporting all the possibilities and especially different ways of messaging is a bit of a pain.

I'll try to get as much progress done and open a draft PR if I can, working on a POC right now

**Comment by 2wheeh** [comment]
FYI, this one looks good as well: https://github.com/StyleT/webext-pegasus
facebook lexical is using it with WXT.

**Comment by spookyuser** [comment]
What's facebook's lexical?

> FYI, this one looks good to reference as well: https://github.com/StyleT/webext-pegasus facebook's lexical is using it with WXT.

**Comment by aklinker1** [comment]
> FYI, this one looks good to reference as well: https://github.com/StyleT/webext-pegasus facebook's lexical is using it with WXT.

Nice, yeah, that's what I'm thinking it would look like. I mentioned having a transport layer abstracted out of the more useful utils like RPC or event broadcasting.

For more details on what we want to implement: https://github.com/wxt-dev/wxt/pull/899#pullrequestreview-2217270397

However, this hasn't been a priority, so we can continue to recommend other messaging frameworks as well.

**Comment by Timeraa** [comment]
Yeah I haven't had the time to continue on it since my extension took up way too much time lately

**Comment by 2wheeh** [comment]
@spookyuser 
> What's facebook's lexical?

It is a RTE framework by meta: https://lexical.dev
They are using WXT on their devtools extension (where I met WXT at first): https://github.com/facebook/lexical/tree/main/packages/lexical-devtools

**Comment by aklinker1** [comment]
Oh, nice, Facebook is using WXT? That's a big win lol

**Comment by Timeraa** [comment]
That is nice indeed

**Comment by aleclarson** [comment]
I'd like to note that a solution needs to account for the lack of `externally_connectable` support on Firefox.

**Comment by aklinker1** [comment]
> I'd like to note that a solution needs to account for the lack of `externally_connectable` support on Firefox.

👀 Didn't know Firefox doesn't support it... Good to know, thanks. The alternatives, window messaging and custom events, should be supported as well anyways.

**Comment by davestewart** [comment]
In my own extensions I wrote/use this:

- https://github.com/davestewart/extension-bus

It makes it super-simple to set up coms in in each process, and target handlers in any other process.

**Comment by maribox** [comment]
Have you ever thought about adding this feature to the v1 roadmap?

**Comment by aklinker1** [comment]
This isn't a blocker for v1, so I won't add it to the milestone. It will be a separate package (or packages), and can be added into WXT at any time if we decide to add them to core.

**Comment by molvqingtai** [comment]
By exposing adapters and allowing users to customize communication methods, [comctx ](https://github.com/molvqingtai/comctx) can fulfill all these requirements. Since rewriting [messaging](https://webext-core.aklinker1.io/messaging/installation/) and [proxy-service](https://webext-core.aklinker1.io/proxy-service/installation/) would require significant changes, I've created a new repository that serves as a superior alternative to proxy-service.

Example: `injected-script ⇆ content-script ⇆ background ⇆ popup-page`
https://github.com/molvqingtai/comctx/tree/master/examples/browser-extension

**Comment by ion1** [comment]
Web workers created by content scripts and injected scripts could be added to the mix.

Perhaps some kind of a hierarchical addressing scheme could be used, and every participant could have a routing table to determine where to forward a message next.

Security needs to be taken into account, too: if an injected script can send messages to the system, so can any code running on the web page in general.

**Comment by ion1** [comment]
As for how to specifically implement the communication between a content script and an injected script, I propose using `CustomEvent`s dispatched on the `script` element created by `injectScript`. The injected script has access to the element through `document.currentScript`.

As the element is unique per script, there are no concerns about the communication with multiple injected scripts getting mixed up. The element being removed from the DOM is not a problem either.

It also requires a bit more effort to detect for the purpose of fingerprinting users than, say, just watching for `message` events on `window`.

**Comment by aklinker1** [comment]
It's been a while, but I've been motivated to try and move forward with this feature. Here's my proposal, it's slightly different from my last comments, inspired by ideas from the different messaging libraries everyone shared. If you're familiar with them, hopefully you can see where my inspirations came from.

The idea behind this hypothetical extension is that it puts netflix player controls and show info in the popup.

<details>
<summary>Messaging Setup and Code</summary>

```ts
// messaging-graph.ts

import {
  defineMessagingGraph,
  type BackgroundNode,
  type PageNode,
  type ContentScriptNode,
  type InjectedScriptNode,
  type MessageSender,
} from 'wxt/messaging';

type ActiveTabChangedEvent = (
  data: chrome.tabs.Tab | undefined,
  sender: Sender,
) => void;

// This could be generated by WXT
const NODE_MAP = {
  background: 'background',
  popup: 'page',
  'netflixPlayer': 'content-script',
  'netflixPlayerInjected': 'main-world-script';
}

// Here is where you define the types for the entire message passing system - the nodes that are present, the RPC functions they implement, and events they want to listen to.
export type MessagingGraph = {
  background: BackgroundNode<{
    rpc: {
      // Define RPC functions the node must implements
      fetchJson: (
        url: string,
        bodyInit?: { headers?: Record<string, string>; body?: any },
      ) => any;
    };
  }>;
  popup: PageNode<{
    events: {
      // Define events the node needs to listen to
      ACTIVE_TAB_CHANGED: ActiveTabChangedEvent;
    };
  }>;
  // Nodes that don't impeent any RPC functions or care about any events may still be needed as "pass-through" nodes that forward messages to the nodes it's connected to.
  netflixPlayer: ContentScriptNode;
  // Some noes need additional info, like the parent content script they communicate with.
  netflixPlayerInjected: MainWorldScriptNode<
  'netflixPlayer',
    {
      rpc: {
        getCurrentShow: () => NetflixMedia;
        play: () => void;
        pause: () => void;
        // ...
      };
      events: {
        ACTIVE_TAB_CHANGED: ActiveTabChangedEvent;
      };
    }
  >;
};

export const messagingGraph = defineMessagingGraph<MessagingGraph>(NODE_MAP);
```

```ts
// background.ts

const rpc: NodeRpc<MessagingGraph, 'background'> = {
  fetchJson: (url, bodyInit) => fetch(url, bodyInit).then((res) => res.json()),
};

const messenger = messagingGraph.initNode('background', {
  // Register RPC implementations during initialization
  rpc,
});

chrome.tabs.onActivated.addListener((tab) => {
  // Broadcast an event to all nodes
  void messenger.broadcast('ACTIVE_TAB_CHANGED', tab);
});
```

```ts
// popup/index.ts

const messenger = messagingGraph.initNode('popup');

messenger.on('ACTIVE_TAB_CHANGED', (tab) => {
  const show = await messenger.netflixPlayerInjected.target(tab.id).getShow();
  updateShow(show);
});
```

```ts
// netflix-player.content.ts

// You need to init pass-through nodes still, even if you don't use the returned messenger in that node.
messagingGraph.initNode("netflixPlayer");
```

```ts
// netflix-player-injected.ts

const rpc: NodeRpc<MessagingGraph, 'netflixPlayerInjected'> = {
  getCurrentShow: () => {
    // ...
  },
  // ...
}

const messenger = messagingGraph.initNode("netflixPlayerInjected", {
  rpc,
  // Pass a transports object to override defaults
  transports: {
    netflixPlayer: createExternalTransport({
      // Options...
    }),
  }
});
```

</details>

In theory, all that should be able to be implemented, though I have not started.

Some comments:

1. Naming is up for debate. I like the theme I chose, naming things "graph" and "node". I did use "transport" instead of "edge", but I think that makes sense in context it's used. But if anyone has any other ideas for naming, I'm down to change it up.
2. Transports would be defined for every way of sending messages in an extension (native, external, custom events, ports, single-messages, custom events, window messaging, etc)
   - Some will have targets, like pointing to a tab/frame for content scripts, others won't like the background.
   - The library would automatically choose transports for most cases, so you would only need to pass them to override the default one chose by the package or pass required config.
3. I struggled quite a bit coming up with a the type and runtime args for `defineMessagingGraph`. In the end, I settled on passing in a basic map so the runtime has enough info to assign default transports, while the type parameter defines types for the full system.
4. Unlike what I said in the issue description, I was able to design a package that all the message passing patterns can use - RPC, broadcasting events, and sending messages to specific targets.
5. Tree-shaking - RPC is fully tree-shakable, implementations aren't bundled in entrypoints that call it from a different context. But transports are not tree-shakable right now. To fix that, you have to define transports in each call to `initNode`, the package couldn't automatically pick one to use. Transports would all be very small... so it wouldn't cause bundles to be much larger if we always included them...

The main thing I'm not happy with is how complex the type definition for the graph is... but I'm kinda lost on how it could be simplified. Otherwise I can't decide between balancing simplicity and bundle size when deciding whether or not to make transports tree-shakable.

Would appreciate feedback on all of this.
